### ISSUE TITLE
Total build time for a given component doesnt seem to match the individual time of the different tasks

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#300]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/300)
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Mar 13, 2018, 18:53

Take a look here for example:

See https://gitlab.gnome.org/GNOME/gnome-build-meta/issues/23#note_80447

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 14, 2018, 10:00

This is actually expected. BuildStream makes a statement in the log that it took *N* time to do a *thing*, and that a complete tasks took *N* time; we make no guarantee that the sum of *timed things* will add up to the total time of the task.

At the very best; if we were to be very very good, we still would not be able to guarantee this sum to be reliable, at least with current APIs in use (because the kernel scheduling could very well put a task to sleep in between two timed activities and wait some time before re-scheduling that process).

Some activities however are completely unaccounted for, the one I can think of most notably is that we dont bother to time and report how much time it takes to clean up (remove) the build directory after having cached an artifact.

As a mere matter of opinion, I left that out because I considered the time to cleanup build directories to be more *"clutter"* in the UI / log than interesting information, however I don't have a strong opinion on this; we can start to time that.

In addition to the above, it should be observed that:
* BuildStream core only cares about the main task time, as we cannot be in control of plugins
* BuildStream provides a `Plugin.timed_activity()` context manager for plugins to mark the time it takes to do things
* Timed activities *can be nested*, although this doesnt normally happen (in the case of deep nesting, one would again not have a sensible *sum*, instead the *sum* of activities would be *greater* than the complete task time)

In short, I'm not opposed to timing more things; but a guarantee that the sum of subtasks equal the time of a full task is not really reasonable afaics, it would require a completely different API for timing than the context manager we have (we would have to have plugins *"mark"* events which occurred instead of letting the plugin time individual things).

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Mar 15, 2018, 18:38

ok, so to put this in perspective I'm going to copy the analisys done in the linked issue here:

This element took 00:24:59 to build:

```
[00:24:59][6c2854e8][build:core-deps/geoclue.bst         ] SUCCESS
```

this splits in

```
[00:11:11][6c2854e8][build:core-deps/geoclue.bst         ] SUCCESS Staging dependencies
[00:05:36][6c2854e8][build:core-deps/geoclue.bst         ] SUCCESS Integrating sandbox
[00:00:00][6c2854e8][build:core-deps/geoclue.bst         ] SUCCESS Staging sources
[00:00:17][6c2854e8][build:core-deps/geoclue.bst         ] SUCCESS Running configure-commands
[00:02:25][6c2854e8][build:core-deps/geoclue.bst         ] SUCCESS Running build-commands
[00:00:00][6c2854e8][build:core-deps/geoclue.bst         ] SUCCESS Running install-commands
[00:00:00][6c2854e8][build:core-deps/geoclue.bst         ] SUCCESS Running strip-commands
[00:01:28][6c2854e8][build:core-deps/geoclue.bst         ] SUCCESS Caching Artifact
```
This is:
- final build time: 00:24:59
- sum of operations: 00:20:57

I would not mind if the difference is a couple of seconds, but we are talking here ~**4** minutes (~**16%** of the total time), which I would like to know where they go

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 16, 2018, 06:35

[[Gitlab user @jjardon]](https://gitlab.com/jjardon) Thanks for the additional feedback.

Do you think it would help if we added some tag for issues, like "Patches welcome" ? I think leaving the issue open is enough to communicate that (but I've seen other projects use that kind of tag, like "Help Needed" and such).

My initial reply was verbose, but I am not opposed to timing more things.

If it gets to be a *lot*, we may need to mitigate more messages with finer grained verbosity control.

Note that for that machine, it took 11min to stage, it doesnt seem completely unreasonable that it might take 4min to remove what it took 11min to stage.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on Jun 21, 2018, 07:31

marked this issue as related to #244

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on Jun 21, 2018, 07:32

removed the relation with #244
